File: App/composeApp/src/commonMain/kotlin/com/lazy/travel/core/network/ApiResult.kt
package com.lazy.travel.core.network

import kotlinx.serialization.Serializable

/**
 * Standard API response — matches server ResultApi exactly
 *
 * Server ResultType enum:
 *   Success   = 1
 *   Exception = 2
 *   Fail      = 3
 */
@Serializable
data class ApiResult<T>(
    val Type: Int = 3,
    val Object: T? = null,
    val Description: String? = null,
    val Total: Double? = null,
    val ObjectExtra: T? = null,
    val Success: Boolean? = null,
) {
    val isSuccess get() = Type == 1
    val isError   get() = Type != 1
    val messageCode get() = Description ?: ""
}

@Serializable
data class PagedResult<T>(
    val Type: Int = 3,
    val Object: List<T> = emptyList(),
    val Description: String? = null,
    val Total: Double? = null,
)  {
    val isSuccess get() = Type == 1
}

/** Sealed result for UI layer */
sealed class LtResult<out T> {
    data class Success<T>(val data: T) : LtResult<T>()
    data class Error(val code: String, val message: String? = null) : LtResult<Nothing>()
    data object Loading : LtResult<Nothing>()
}

File: App/lib/core/network/api_result.dart
sealed class ApiResult<T> {
  const ApiResult();
}

class ApiSuccess<T> extends ApiResult<T> {
  final T data;
  const ApiSuccess(this.data);
}

class ApiError<T> extends ApiResult<T> {
  final String message;
  final int? statusCode;
  const ApiError(this.message, {this.statusCode});
}

extension ApiResultX<T> on ApiResult<T> {
  T? get dataOrNull => this is ApiSuccess<T> ? (this as ApiSuccess<T>).data : null;
  String? get errorOrNull => this is ApiError<T> ? (this as ApiError<T>).message : null;
  bool get isSuccess => this is ApiSuccess<T>;
  bool get isError => this is ApiError<T>;

  R when<R>({
    required R Function(T data) success,
    required R Function(String message, int? statusCode) error,
  }) {
    return switch (this) {
      ApiSuccess<T> s => success(s.data),
      ApiError<T> e   => error(e.message, e.statusCode),
    };
  }
}
